pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go

```diff
@@ -0,0 +1,477 @@
+	}
+	if isExported, err := r.isServiceExported(
+		ctx,
+		types.NamespacedName{Name: serviceName, Namespace: localEpSlice.Namespace},
+	); !isExported || err != nil {
+		return false, err
+	}
+	// Only mirrors EndpointSlice compatible with the derived Service IP family
+	if !slices.Contains(derivedService.Spec.IPFamilies, corev1.IPFamily(localEpSlice.AddressType)) {
+		return false, nil
+	}
+	return true, nil
+}
+
+func getDerivedServiceName(localEpSlice *discoveryv1.EndpointSlice) string {
+	serviceName := localEpSlice.Labels[discoveryv1.LabelServiceName]
+	if serviceName == "" {
+		return ""
+	}
+	return derivedName(types.NamespacedName{
+		Name: serviceName, Namespace: localEpSlice.Namespace,
+	})
+}
+
+func getLocalDerivedEndpointSliceName(localEpSlice *discoveryv1.EndpointSlice) string {
+	suffix := getSuffix(localEpSlice)
+	derivedServiceName := getDerivedServiceName(localEpSlice)
+	if derivedServiceName == "" || suffix == "" {
+		return ""
+	}
+	return derivedServiceName + "-" + suffix
+}
+
+func (r *mcsAPIEndpointSliceMirrorReconciler) getLocalDerivedEndpointSlice(
+	ctx context.Context, localEpSlice *discoveryv1.EndpointSlice,
+) (*discoveryv1.EndpointSlice, error) {
+	derivedEpSliceName := getLocalDerivedEndpointSliceName(localEpSlice)
+	if derivedEpSliceName == "" {
+		return nil, nil
+	}
+	var derivedEpSlice discoveryv1.EndpointSlice
+	if err := r.Client.Get(
+		ctx,
+		types.NamespacedName{
+			Name:      derivedEpSliceName,
+			Namespace: localEpSlice.Namespace,
+		},
+		&derivedEpSlice,
+	); err != nil {
+		return nil, client.IgnoreNotFound(err)
+	}
+	return &derivedEpSlice, nil
+}
+
+func (r *mcsAPIEndpointSliceMirrorReconciler) updateDerivedEndpointSlice(
+	derivedEpSlice, localEpSlice *discoveryv1.EndpointSlice, derivedService *corev1.Service,
+) {
+	controllerutil.SetControllerReference(derivedService, derivedEpSlice, r.Scheme())
+
+	derivedEpSlice.Labels = maps.Clone(derivedService.Labels)
+	if derivedEpSlice.Labels == nil {
+		derivedEpSlice.Labels = map[string]string{}
+	}
+	derivedEpSlice.Labels[mcsapiv1alpha1.LabelServiceName] = localEpSlice.Labels[discoveryv1.LabelServiceName]
+	derivedEpSlice.Labels[discoveryv1.LabelServiceName] = derivedService.Name
+	derivedEpSlice.Labels[mcsapiv1alpha1.LabelSourceCluster] = r.clusterName
+	derivedEpSlice.Labels[discoveryv1.LabelManagedBy] = endpointSliceLocalMCSAPIControllerName
+
+	derivedEpSlice.Annotations = nil
+	derivedEpSlice.AddressType = localEpSlice.AddressType
+
+	// Beware those are shallow copies, content of the struct should not be modified
+	derivedEpSlice.Endpoints = slices.Clone(localEpSlice.Endpoints)
+	derivedEpSlice.Ports = shallowCopyServicePortToEndpointPort(derivedService.Spec.Ports)
+}
+
+func shallowCopyServicePortToEndpointPort(ports []corev1.ServicePort) []discoveryv1.EndpointPort {
+	endpointPorts := make([]discoveryv1.EndpointPort, 0, len(ports))
+	for _, port := range ports {
+		endpointPort := discoveryv1.EndpointPort{
+			Port:        &port.Port,
```

Comment:
This should be the target port, not port.

Reply:
Yeah, preserving the ones from the local epslice seems the easiest to avoid having to reimplement a lot of conversion logic. I'm not exactly sure if there could be cases of conflicts that cause a different subset of ports to be chosen by the service (same name but different port maybe?), but handling those is probably out of scope here, at least for the moment.

Reply:
Hmm yep also I think most of those weird cases should be caught by a ServiceExport conflict, then the behavior might not be the best on port conflict but at least there is some kind of warning exposed to tell the user that there is something wrong...